File: Week_03/java/src/_93_RestoreIPAddresses.java
import java.util.ArrayList;
import java.util.List;

/**
 * 第一遍：2020/08/09周日 ✅
 * 第二遍：2020/08/09周日
 * 第三遍：2020/07/18周六
 * 第三遍：2020/07/08周四
 * 自己做出来了，但是是枚举的三种情况。。。
 */
class _93_RestoreIPAddresses {
    public List<String> restoreIpAddresses(String s) {
        List<String> ret = new ArrayList<>();
        dfs(s, 0, 4, new StringBuilder(), ret);
        return ret;
    }

    void dfs(String s, int l, int count, StringBuilder sb, List<String> ret) {
        if (l == s.length() && count == 0) {
            sb.deleteCharAt(sb.length() - 1);
            ret.add(sb.toString());
            sb.append('.');
            return;
        }

        if (count == 0) {
            return;
        }

        for (int i = 1; i < 4; i++) {
            if (l + i > s.length() || (i > 1 && s.charAt(l) == '0')
                    || (i == 3 && (s.charAt(l) - '0') * 100 + (s.charAt(l + 1) - '0') * 10 + (s.charAt(l + 2) - '0') > 0xFF)) {
                break;
            }

            sb.append(s, l, l + i).append('.');
            dfs(s, l + i, count - 1, sb, ret);
            sb.delete(sb.length() - i - 1, sb.length());
        }
    }
}

File: Week_03/kotlin/src/_93_RestoreIPAddresses.kt
fun restoreIpAddresses(s: String): List<String> {
    fun String.dfs(segStart: Int, segments: IntArray, segId: Int, ret: ArrayList<String>) {
        if (segStart == length && segId == 4) {
            val sb = StringBuilder()
            for (seg in segments) {
                sb.append(seg).append('.')
            }
            sb.deleteCharAt(sb.length - 1)
            ret.add(sb.toString())
        }
        if (segId == 4) {
            return
        }

        for (i in 1..4) {
            if (segStart + i > length || (i > 1 && (this[segStart] == '0'))) {
                continue
            }

            val seg = Integer.parseInt(substring(segStart, segStart + i))
            if (seg > 0xFF) {
                continue
            }
            segments[segId] = seg
            dfs(segStart + i, segments, segId + 1, ret);
        }
    }

    val segments = IntArray(4)
    val ret = ArrayList<String>()
    s.dfs(0, segments, 0, ret)
    return ret
}